### PR TITLE
Hierarchy Additions

### PR DESCRIPTION
**Pull request by pzalep1**
This PR adds a builder for hierarchy building for admins and editors.
Screen Shot 2022-08-04 at 9 54 39 AM
Screen Shot 2022-08-04 at 9 55 20 AM
Screen Shot 2022-08-04 at 9 54 51 AM

### CODE REVIEW

File: src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  selector: 'clark-hierarchy-builder',
+  templateUrl: 'hierarchy-builder.component.html',
+  styleUrls: ['hierarchy-builder.component.scss'],
+})
+export class HierarchyBuilderComponent implements OnInit {
+
+  // This is just dummy data that is replaced in the OnInit function with the parent data
+  TREE_DATA: LearningObjectNode[] = [
+    {
+      "name": "paige",
```
Comment (mfranz1) [inline src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts:19]:
```suggestion
      "name": "paige the rage",
```
Gotta leave your mark on the team

File: src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    this.dataSource.remove(node);
+  }
+
+  checkBox() {
+    this.acknowledge = !this.acknowledge;
+  }
+
+  validateSubmitable(node) {
+    if(node.name.length <= 2) {
+      // Pop up an error
```
Comment (mfranz1) [inline src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts:77]: You gonna pop the error?

File: src/app/admin/core/collection.service.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import { HttpClient } from '@angular/common/http';
 import { Injectable } from '@angular/core';
-import { ADMIN_ROUTES } from '@env/route';
+import { ADMIN_ROUTES} from '@env/route';
```
Comment (mfranz1) [inline src/app/admin/core/collection.service.ts:3]:
```suggestion
import { ADMIN_ROUTES } from '@env/route';
```

File: src/app/admin/pages/learning-objects/learning-objects.component.ts [hunk 4]
```diff
@@ -91,10 +91,10 @@ export class LearningObjectsComponent
...
     });
 
-    this.searchService.needsChange$.subscribe( () => {
-      this.learningObjects = [];
-      this.getLearningObjects();
-    });
+    // this.searchService.needsChange$.subscribe( () => {
+    //   this.learningObjects = [];
+    //   this.getLearningObjects();
+    // });
```
Comment (mfranz1) [inline src/app/admin/pages/learning-objects/learning-objects.component.ts:97]: Let's remove this

**Review by mfranz1 — CHANGES_REQUESTED** [review]
Can you lint your client?

File: src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    this.TREE_DATA[0] = { _id: this.parent.id, name: this.parent.name, length: this.parent.length, showForm: false, children: []};
+    this.contributors = this.parent.contributors.map(contrib => {
+      return contrib.username
+    });
+  }
+
+  hasChild = (_: number, node: LearningObjectNode) =>
+    !!node.children && node.children.length > 0;
+
+  addLearningObject(parentNode: LearningObjectNode) {
```
Comment (EVogel1999) [inline src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts:47]: The functions in this file need block comments and inline comments.

File: src/app/admin/components/hierarchy-builder/hierarchy-object/hierarchy-object.component.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    }
+  }
+
+
+  remove() {
+    this.removeLo.emit(this.node);
+  }
+
+
+  forbiddenNameValidator(): AsyncValidatorFn {
```
Comment (EVogel1999) [inline src/app/admin/components/hierarchy-builder/hierarchy-object/hierarchy-object.component.ts:98]: The functions in this file need block and inline comments.

File: src/app/admin/components/hierarchy-builder/tree-datasource.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    });
+
+    if (updatedTree!) {
+      tree.children![updatedIndex!] = updatedTree!;
+      return true;
+    }
+    return false;
+  }
+
+  moveExpansionState(from: LearningObjectNode, to: LearningObjectNode) {
```
Comment (EVogel1999) [inline src/app/admin/components/hierarchy-builder/tree-datasource.ts:93]: The functions in this file need block and inline comments.

File: src/app/admin/components/learning-object-list-item/learning-object-list-item.component.ts [hunk 8]
```diff
@@ -187,6 +190,26 @@ export class LearningObjectListItemComponent implements OnChanges {
     });
   }
 
+  async checkForChildren() {
```
Comment (EVogel1999) [inline src/app/admin/components/learning-object-list-item/learning-object-list-item.component.ts:193]: Comments.

File: src/app/admin/components/learning-object-list-item/learning-object-list-item.component.ts [hunk 9]
```diff
@@ -216,6 +239,16 @@ export class LearningObjectListItemComponent implements OnChanges {
    }
 
 
+   openHierarchy(learningObject: any) {
+     console.log(this.showHierarchyBuilder)
```
Comment (EVogel1999) [inline src/app/admin/components/learning-object-list-item/learning-object-list-item.component.ts:243]: 🔥 console log

File: src/app/admin/core/hierarchy.service.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        children
+      },
+      {
+        withCredentials: true,
+        responseType: 'text'
+      }
+    ).toPromise();
+  }
+
+  async checkName(username: string, objectName: string): Promise<boolean> {
```
Comment (EVogel1999) [inline src/app/admin/core/hierarchy.service.ts:35]: Comments.

**Review by EVogel1999 — COMMENTED** [review]
Mostly just comments.

File: src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts (hunk 2 above)
Comment (pzalep1) [reply src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts:77]: I moved this to the component with a mat-error instead I'll remove it here.

File: src/app/admin/pages/learning-objects/learning-objects.component.ts [hunk 11]
```diff
@@ -181,6 +176,7 @@ export class LearningObjectsComponent
         // this is a false event from the virtual scroller, disregard
         return;
       }
+      console.log(this.learningObjects.length)
```
Comment (mfranz1) [inline src/app/admin/pages/learning-objects/learning-objects.component.ts:179]:
```suggestion
```
yeet

**Review by mfranz1 — APPROVED** [review]
Looks good! just a console log

File: src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    }
+    const newNode = await this.setParents(parentId, childrenIds);
+    if(node.name === this.parent.name) {
+      this.toaster.success(
+        'Success',
+        'Hierarchy Created!'
+      );
+      this.close.emit();
+    }
+    return newNode;
```
Comment (EVogel1999) [inline src/app/admin/components/hierarchy-builder/hierarchy-builder.component.ts:107]: Could we add some more inline comments for this function?

**Review by EVogel1999 — COMMENTED** [review]
Just one suggestion to add more inline comments for what some of the recursive functions are doing.  Looks good after that!

**Review by EVogel1999 — APPROVED** [review]

**Pull request merged**